### PR TITLE
Multi level supported

### PR DESCRIPTION
**Pull request by jackyhuang85**

### CODE REVIEW

File: src/main_cache.hpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#include "base_cache.hpp"
+#include <random>
+
+class MainCache : public BaseCache {
+  public:
+    MainCache(const CACHE_SET &);
```
Comment (jerryzj) [inline src/main_cache.hpp]: Codacy Issue found: [Class 'MainCache' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/manual/jerry0715no14/Cache_sim/pullRequest?prid=4735708)

**Comment by jerryzj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/jerry0715no14/Cache_sim/pullRequest?prid=4735708&bid=15809161)

File: src/simulator.cpp [hunk 2]
```diff
@@ -33,118 +38,140 @@ void Simulator::RunSimulation() {
...
+                ++_counter.store_hit;
+                ++_counter.hit_in_main;
+            } else {
+                _is_hit = false;
+                _cache->Set(addr);
+            }
+        }
+    } else {
+        if (_cache_hierarchy_list[0].Get(addr)) {
+            _is_hit = true;
```
Comment (jerryzj) [inline src/simulator.cpp]: Codacy Issue found: [Variable '_is_hit' is assigned a value that is never used.](https://app.codacy.com/manual/jerry0715no14/Cache_sim/pullRequest?prid=4735708)

File: src/simulator.cpp [hunk 3]
```diff
@@ -33,118 +38,140 @@ void Simulator::RunSimulation() {
...
             ++_counter.store_hit;
             ++_counter.hit_in_main;
-
         } else {
-            this->main_cache->_Read(next_addr);
+            _is_hit = false;
```
Comment (jerryzj) [inline src/simulator.cpp]: Codacy Issue found: [Variable '_is_hit' is assigned a value that is never used.](https://app.codacy.com/manual/jerry0715no14/Cache_sim/pullRequest?prid=4735708)

File: src/base_cache.hpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
+#ifndef BASECACHE_HPP
+#define BASECACHE_HPP
+
+#include "cache_setting.hpp"
+#include "datatype.hpp"
+#include <cmath>
+#define MAX_LINE 65536
+
+struct CachePropertyStruct {
```
Comment (jerryzj) [inline src/base_cache.hpp]: `struct CachePropertyStruct ` holds many content which `struct CACHE_SET` already have.
I think there is no need to create another structure

**Review by jerryzj — CHANGES_REQUESTED** [review]
The use of similar structure should be avoided.

File: src/main_cache.cpp [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    _Replace(addr);
+    return true;
+}
+
+bool MainCache::IsHit(const addr_t &addr) {
+    ulint idx(0);
+    bool identical(true);
+
+    auto check_ident = [&](const ulint idx, const uint _bit_tag,
+                           const addr_t lhs, const addr_t rhs[]) -> bool {
```
Comment (jerryzj) [inline src/main_cache.cpp]: Codacy Issue found: [Function parameter 'lhs' should be passed by const reference.](https://app.codacy.com/manual/jerry0715no14/Cache_sim/pullRequest?prid=4735708)

**Comment by jerryzj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/jerry0715no14/Cache_sim/pullRequest?prid=4735708&bid=15809161)

**Comment by jerryzj** [comment]
I also use lambda function to reduce repetitive codes in function `bool MainCache::IsHit`

File: src/base_cache.cpp [hunk 6]
```diff
@@ -0,0 +1,8 @@
+#include "base_cache.hpp"
+
+BaseCache::BaseCache(const CachePropertyStruct &setting) {
+    property = setting;
```
Comment (jerryzj) [inline src/base_cache.cpp]: Codacy Issue found: [Variable 'property' is assigned in constructor body. Consider performing initialization in initialization list.](https://app.codacy.com/manual/jerry0715no14/Cache_sim/pullRequest?prid=4735708)

**Comment by jerryzj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/jerry0715no14/Cache_sim/pullRequest?prid=4735708&bid=15809161)

File: src/base_cache.cpp (hunk 6 above)
Comment (jackyhuang85) [reply src/base_cache.cpp]: oops

**Pull request merged**